### PR TITLE
[Console] Add method to "clear the screen"

### PR DESCRIPTION
**Pull request by alexbowers**
| Q             | A
| ------------- | ---
| Branch?       | 3.4 
| Bug fix?      | no
| New feature?  | yes
| BC breaks?    | yes - Changes interface
| Tests Pass? | yes
| Deprecations? | no 
| License       | MIT
| Doc PR        | symfony/symfony-docs#...  -- not yet

This allows the user to clear the screen of all existing output.

### CODE REVIEW

**Review by javiereguiluz — COMMENTED** [review]
I've needed this sometimes in the past. I'm 👍 .

I have two questions though:

* Are we 100% sure that this works on 100% of operating systems? (I know we have a lot of issues with the console in windows)
* Is the method name clear enough? Maybe `clearScreen()` is a better name? (I know the command is called `clear` (`cls` on Windows) ... but in this context, "clear" can be a lot of things.

Thanks!

File: src/Symfony/Component/Console/Output/OutputInterface.php [hunk 1]
```diff
@@ -48,6 +48,11 @@ public function write($messages, $newline = false, $options = 0);
     public function writeln($messages, $options = 0);
 
     /**
+     * Clears the screen of output.
+     */
+    public function clear();
```
Comment (stof) [inline src/Symfony/Component/Console/Output/OutputInterface.php]: Adding new methods in the interface is a BC break (btw, you haven not even updated all implementations in the component, so the component itself is affected by the BC break)

**Review by stof — CHANGES_REQUESTED** [review]

**Comment by alexbowers** [comment]
@javiereguiluz I am happy for a name change. I'll leave this up to whomever wishes to review. Feel free to suggest changes that are better.

In my view, `clear` is best, because that represents the command as it would be ran in CLI.

In terms of windows, I'm afraid I have absolutely no idea if it works.

I can't imagine there are any ill-effects if it doesn't work, since it would just print the characters to the screen.

I do not have a windows machine to test this against, so cannot verify it.

if anyone has a windows machine and is able to let me know if this:

- Works as expected
- Does not work at all
- Does nothing
- Does something unexpected

That'd be great.

---

@stof Since this is a BC change, does that mean it has to aim for 4.0 instead of 3.4?

Alternatively, is adding this to the interface something that should be done?

Since this isn't a _required_ function as such, for any / all implementations, would this be best resolved by removing the interface method?

--

I'll work on fixing this for all implementations in the component now, I actually wrote this without running the tests locally in a quick push fashion.

I'll work on improving this..

Thanks for the feedback.

File: src/Symfony/Component/Console/Output/OutputInterface.php (hunk 1 above)
Comment (alexbowers) [inline src/Symfony/Component/Console/Output/OutputInterface.php]: @stof This has now been fixed. Tests pass again.

My bad :)

**Comment by chalasr** [comment]
Even for 4.0, we can't break BC without proper upgrade path. What could be done is to introduce a `ClearableOutputInterface` implemented by `Output` subclasses and typehint againt this interface in userland.

**Comment by alexbowers** [comment]
How would this look compared to the code I have put in the file?

Would the existing implementations just implement the interface as well, and that'd be fine?

is the issue that I cannot change the existing interface without BC?

**Comment by chalasr** [comment]
> Would the existing implementations just implement the interface as well, and that'd be fine?

yes. I would let the abstract `Output` unchanged also, updating only concrete implementations (just my own opinion).

> is the issue that I cannot change the existing interface without BC?

yes, changing the existing interface breaks BC as it prevents any app with an implementation from upgrading (running in a fatal error because of the missing method).

File: src/Symfony/Component/Console/Output/OutputInterface.php (hunk 1 above)
Comment (Taluu) [inline src/Symfony/Component/Console/Output/OutputInterface.php]: Still a BC break, as you added a method to the interface. It should be added only on 4.0. Or make a new temporary interface for 3.4.

**Comment by nicolas-grekas** [comment]
In terms of architecture, I'm not sure this belongs to the Output class. The TTY concept is not know by it. Maybe OutputFormatter is a better fit? That would match the "isDecorated" state, so that the `clear()` would be a no-op on non decorated streams.

**Comment by chalasr** [comment]
@nicolas-grekas OutputFormatter does not know about TTY, only [StreamOutput does](https://github.com/symfony/symfony/blob/master/src/Symfony/Component/Console/Output/StreamOutput.php#L93) and flags itself as decorated accordingly (and flags the formatter)

**Review by keradus — CHANGES_REQUESTED** [review]
except of already raised concerns...

you have added new feature, yet no tests for it. add new tests.

**Comment by alexbowers** [comment]
@keradus How would you test that the screen has been cleared?

I could test that the "output" is that string, is that sufficient?

**Comment by keradus** [comment]
probably

**Comment by chalasr** [comment]
@alexbowers Would you mind to finish this PR?

What remains to do for me is to move the TTY specific `Output::clear()` to `StreamOutput` (should it throw if not decorated?) and add a test for it as suggested. 
Given that the new method would be a noop for other output subclasses, I personally don't feel the need for an interface right now (just check that `Output` is an instance of `ConsoleOutput` or `StreamOutput`, as for writing to stderr), and we can still add one later if needed.

**Comment by alexbowers** [comment]
I'll try to get around to it this evening, but can't promise.

I'm happy for somebody else to take this and PR it in a suitable way, if anybody feels like it.

**Comment by nicolas-grekas** [comment]
@alexbowers would you have some time to finish this PR? If not, should we close it?

**Comment by alexbowers** [comment]
Ah, i always forget my PRs.

I still want it to be completed, so I will work on it at some point.
Hopefully, soon.

Sorry for the delay.

**Comment by alexbowers** [comment]
@nicolas-grekas can this be closed in favour of #24449 ?

**Pull request closed without merging**